### PR TITLE
[libftdi1] Fix, cleanup, C++ binding, test port

### PR DESCRIPTION
**Pull request by dg0yt**
Fixes installed cmake config and pc files.
Resolves #40846 (libftdipp1).

### CODE REVIEW

File: ports/libftdi1/disable-config-script.diff [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+--- a/CMakeLists.txt
++++ b/CMakeLists.txt
+@@ -170,6 +170,7 @@ install(FILES ${CMAKE_BINARY_DIR}/libftdi1.pc ${CMAKE_BINARY_DIR}/libftdipp1.pc
+         DESTINATION lib${LIB_SUFFIX}/pkgconfig)
+ 
+ if (UNIX OR MINGW)
++elseif (0)
+   configure_file ( libftdi1-config.in ${CMAKE_CURRENT_BINARY_DIR}/libftdi1-config @ONLY )
+   install ( PROGRAMS ${CMAKE_CURRENT_BINARY_DIR}/libftdi1-config
+             DESTINATION bin )
```
Comment (dg0yt) [inline ports/libftdi1/disable-config-script.diff:12]: Note: The port always removed the installed script by removing `bin`. The script is unusable without modifications.

File: ports/libftdi1/libftdipp1.diff [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+diff --git a/libftdipp1.pc.in b/libftdipp1.pc.in
+index 8158f38..b4fff0a 100644
+--- a/libftdipp1.pc.in
++++ b/libftdipp1.pc.in
+@@ -8,4 +8,4 @@ Description: C++ wrapper for libftdi1
+ Requires: libftdi1
+ Version: @VERSION@
+ Libs: -L${libdir} -lftdipp1
+-Cflags: -I${includedir}
++Cflags: -I${includedir} -I${prefix}/include
```
Comment (dg0yt) [inline ports/libftdi1/libftdipp1.diff:39]: For boost headers.

File: ports/libftdi1/portfile.cmake [hunk 3]
```diff
@@ -6,44 +6,74 @@ vcpkg_download_distfile(ARCHIVE
...
-if(VCPKG_LIBRARY_LINKAGE STREQUAL "static")
-    file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/bin" "${CURRENT_PACKAGES_DIR}/debug/bin")
+if(NOT "cpp" IN_LIST FEATURES)
+    file(REMOVE  "${CURRENT_PACKAGES_DIR}/lib/pkgconfig/libftdipp1.pc" "${CURRENT_PACKAGES_DIR}/debug/lib/pkgconfig/libftdipp1.pc")
 endif()
 
-vcpkg_copy_pdbs()
+set(file_list "${SOURCE_PATH}/COPYING.LIB")
+if("cpp" IN_LIST FEATURES)
+    set(file_list "${SOURCE_PATH}/LICENSE" "${SOURCE_PATH}/COPYING.LIB" "${SOURCE_PATH}/COPYING.GPL")
```
Comment (WangWeiLin-MV) [inline ports/libftdi1/portfile.cmake:76]:
```suggestion
    list(APPEND file_list "${SOURCE_PATH}/LICENSE" "${SOURCE_PATH}/COPYING.GPL")
```

File: ports/libftdi1/portfile.cmake [hunk 4]
```diff
@@ -6,44 +6,74 @@ vcpkg_download_distfile(ARCHIVE
...
+    vcpkg_replace_string("${CURRENT_PACKAGES_DIR}/share/${PORT}/LibFTDI1Config-debug.cmake" "{VCPKG_IMPORT_PREFIX}/debug/include/" "{VCPKG_IMPORT_PREFIX}/include/")
+    file(READ "${CURRENT_PACKAGES_DIR}/share/${PORT}/LibFTDI1Config.cmake" release_config)
+    file(WRITE "${CURRENT_PACKAGES_DIR}/share/${PORT}/LibFTDI1Config.cmake" "
+if(NOT DEFINED CMAKE_BUILD_TYPE OR CMAKE_BUILD_TYPE MATCHES \"^[Dd][Ee][Bb][Uu][Gg]\$\")
+    include(\"\${CMAKE_CURRENT_LIST_DIR}/LibFTDI1Config-debug.cmake\")
+    return()
+endif()
+${release_config}"
+    )
+endif()
```
Comment (WangWeiLin-MV) [inline ports/libftdi1/portfile.cmake:67]: Just a suggestion, maybe change the release `LibFTDI1Config.cmake` to `LibFTDI1Config-release.cmake`.

File: ports/libftdi1/portfile.cmake [hunk 5]
```diff
@@ -6,44 +6,74 @@ vcpkg_download_distfile(ARCHIVE
...
+    file(RENAME "${CURRENT_PACKAGES_DIR}/debug/lib/cmake/libftdi1/LibFTDI1Config.cmake" "${CURRENT_PACKAGES_DIR}/debug/lib/cmake/libftdi1/LibFTDI1Config-debug.cmake")
+endif()
 vcpkg_cmake_config_fixup(CONFIG_PATH lib/cmake/libftdi1)
-
-vcpkg_fixup_pkgconfig()
+vcpkg_replace_string("${CURRENT_PACKAGES_DIR}/share/${PORT}/LibFTDI1Config.cmake" "/lib/cmake/${PORT}/" "/share/${PORT}/")
+if(NOT VCPKG_BUILD_TYPE)
+    vcpkg_replace_string("${CURRENT_PACKAGES_DIR}/share/${PORT}/LibFTDI1Config-debug.cmake" "/debug/lib/cmake/${PORT}/" "/share/${PORT}/")
+    vcpkg_replace_string("${CURRENT_PACKAGES_DIR}/share/${PORT}/LibFTDI1Config-debug.cmake" "{_IMPORT_PREFIX}" "{VCPKG_IMPORT_PREFIX}")
+    vcpkg_replace_string("${CURRENT_PACKAGES_DIR}/share/${PORT}/LibFTDI1Config-debug.cmake" "{VCPKG_IMPORT_PREFIX}/debug/include/" "{VCPKG_IMPORT_PREFIX}/include/")
```
Comment (WangWeiLin-MV) [inline ports/libftdi1/portfile.cmake:58]: If it is possible to not change this variable? Or at least add `_` to indicate it is non-public variables.

File: ports/libftdi1/portfile.cmake [hunk 6]
```diff
@@ -6,44 +6,74 @@ vcpkg_download_distfile(ARCHIVE
...
+    file(READ "${CURRENT_PACKAGES_DIR}/share/${PORT}/LibFTDI1Config.cmake" release_config)
+    file(WRITE "${CURRENT_PACKAGES_DIR}/share/${PORT}/LibFTDI1Config.cmake" "
+if(NOT DEFINED CMAKE_BUILD_TYPE OR CMAKE_BUILD_TYPE MATCHES \"^[Dd][Ee][Bb][Uu][Gg]\$\")
+    include(\"\${CMAKE_CURRENT_LIST_DIR}/LibFTDI1Config-debug.cmake\")
+    return()
+endif()
+${release_config}"
+    )
+endif()
```
Comment (WangWeiLin-MV) [inline ports/libftdi1/portfile.cmake:68]: Usage file.

```cmake
libftdi1 provides pkg-config modules:

    # Library to program and control the FTDI USB controller
    libftdi1

    # C++ wrapper for libftdi1
    libftdipp1

libftdi1 provides CMake integrations:

    find_package(LibFTDI1 CONFIG REQUIRED)
    target_compile_definitions(main PRIVATE ${LIBFTDI_DEFINITIONS})
    target_include_directories(main PRIVATE ${LIBFTDI_INCLUDE_DIRS})
    target_link_directories(main PRIVATE ${LIBFTDI_LIBRARY_DIRS})

    # Library to program and control the FTDI USB controller
    target_link_libraries(main PRIVATE ${LIBFTDI_LIBRARIES})
    # C++ wrapper for libftdi1
    target_link_libraries(main PRIVATE ${LIBFTDIPP_LIBRARIES})
```

**Review by WangWeiLin-MV — COMMENTED** [review]
The port `libftdi1` usage tests pass with the following triplets:

* x64-windows

File: ports/libftdi1/cmake-version.diff
[the remark this answers is not in the document]
Comment (dg0yt) [reply ports/libftdi1/cmake-version.diff:1]: Independent decision. `cmake_minimum_required` must come before `project`, and <3.5 is causing warning in current versions.

File: ports/libftdi1/portfile.cmake (hunk 5 above)
Comment (dg0yt) [reply ports/libftdi1/portfile.cmake:58]: Both _IMPORT_PREFIX and VCPKG_IMPORT_PREFIX are injected by `vcpkg_cmake_config_fixup`. But that function is tuned for CMake config exported from CMake, and this CMake config is manual stuff. So adapting for vcpkg dual-config had to be solved in a custom way.

File: ports/libftdi1/portfile.cmake (hunk 4 above)
Comment (dg0yt) [reply ports/libftdi1/portfile.cmake:67]: Just isn't enough.

File: ports/libftdi1/portfile.cmake (hunk 6 above)
Comment (dg0yt) [reply ports/libftdi1/portfile.cmake:68]: No, thanks. KISS, accept the output from the tool heuristics.
- The pp variant only exists when the cpp feature is enabled.
- The CMake config is not based on modern export.
- The CMake usage seems to be based on including a usage file. See test port.

File: ports/libftdi1/portfile.cmake (hunk 3 above)
Comment (dg0yt) [reply ports/libftdi1/portfile.cmake:76]: No, thanks. The order is deliberately chosen, APPEND can't handle it.

**Review by dg0yt — COMMENTED** [review]
None of the suggestions accepted.

**Review by WangWeiLin-MV — APPROVED** [review]
The port `libftdi1` usage tests pass with the following triplets:

* x64-linux
* x64-windows

File: ports/libftdi1/linkage.diff [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
+index 17b3617..087fba8 100644
+--- a/src/CMakeLists.txt
++++ b/src/CMakeLists.txt
+@@ -19,8 +19,12 @@ configure_file(ftdi_version_i.h.in "${CMAKE_CURRENT_BINARY_DIR}/ftdi_version_i.h
+ 
+ # Targets
+ set(c_sources     ${CMAKE_CURRENT_SOURCE_DIR}/ftdi.c ${CMAKE_CURRENT_SOURCE_DIR}/ftdi_stream.c CACHE INTERNAL "List of c sources" )
++if(WIN32)
++  set(c_sources     ${CMAKE_CURRENT_SOURCE_DIR}/ftdi.c ${CMAKE_CURRENT_SOURCE_DIR}/exports.def CACHE INTERNAL "List of c sources" )
```
Comment (BillyONeal) [inline ports/libftdi1/linkage.diff:34]: Why no ${CMAKE_CURRENT_SOURCE_DIR}/ftdi_stream.c ?

EDIT: I see this is from the previous win32.patch, no change requested.

**Review by BillyONeal — APPROVED** [review]

**Comment by BillyONeal** [comment]
Thanks!

File: ports/libftdi1/linkage.diff (hunk 7 above)
Comment (dg0yt) [reply ports/libftdi1/linkage.diff:34]: And I tested it already: It doesn't compile on Windows.

**Pull request merged**